Fix Onion v2 support for Neutrino backends

guggero opened the pull request:
Fixes https://github.com/lightningnetwork/lnd/issues/4803.

With https://github.com/lightningnetwork/lnd/pull/737 we intended to add Tor support for Neutrino backends.
Either something changed since that PR or we never fully tested this. Before this fix it was not possible to use an Onion v2 address for the `--neutrino.connect` flag.

With this PR we make it possible to use an Onion v2 hidden service
address as the Neutrino backend.
This failed before because an .onion address cannot be looked up and
converted into an IP address through the normal DNS resolving process,
even when using a Tor socks proxy.
Instead, we turn any v2 .onion address into a fake IPv6 representation
before giving it to Neutrino's address manager and turn it back into an
Onion host address when actually dialing.

To enable Onion v3 support, we need to rewrite the `github.com/btcsuite/btcd/addrmgr` library to be able to use `net.Addr` instead of only `net.IP`. Or we need to rewrite Neutrino to use its own address manager.

bitcoinlizard commented:
I can confirm that this patch resolved my reported issue: 
https://github.com/lightningnetwork/lnd/issues/4803

Roasbeef commented:
> To enable Onion v3 support, we need to rewrite the github.com/btcsuite/btcd/addrmgr library to be able to use net.Addr instead of only net.IP

From the PoV of the Bitcoin p2p network this also requires a series of new addr messages as well. I think the required extensions for this will start being rolled out with the latest release of bitcoind.

Crypt-iQ commented:
Linter complaining. Is it possible for a legitimate ipv6 address to have the "onion" prefix?

guggero commented:
Fixed the linter.

> Is it possible for a legitimate ipv6 address to have the "onion" prefix?

Good question. I added a comment to clarify. But basically that address range can be used to randomly create global unique IDs for local unicast. So these addresses should never be routed to the internet anyway. And even if it would collide with a local interface IP (which is extremely unlikely given the address length), we never actually _dial_ the address. We convert it back to an .onion host name before actually dialing it.

cfromknecht commented on tor/tor_test.go:
```diff
@@ -0,0 +1,32 @@
+	testFakeIP = "fd87:d87e:eb43:58f9:f82e:3e3e:83f3:c625"
+)
+
+func TestOnionHostToFakeIP(t *testing.T) {
```
nit: missing godoc comments on test functions

guggero replied:
Fixed.

cfromknecht reviewed (approved):
nice catch! LGTM 🎉
